### PR TITLE
add Map combinator for the Map type

### PR DESCRIPTION
**Pull request by ildus**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):

* New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

* Add Map combinator for `Map` type.
* Rename old `sum-, min-, max- Map` for mapped arrays to `sum-, min-, max- MappedArrays`

Detailed description / Documentation draft:

Basically this patch gives *Map functions to Map type, and moves old *Map functions to *MappedArrays. But when sumMap, minMap and maxMap called with array arguments they are redirected to *MappedArrays functions.

`sumMapWithOverflow`, `sumMapFiltered` and `sumMapFilteredWithOverflow` keep old name.

I thought about implementing a Map combinator for arrays too, but it seems to be nontrivial task, since sumMap functions for arrays can support various features, like different types for arrays, which is not suitable for combinators where only one nested function is present. Also there are problems with compaction of default values, which requires getting the value from the nested function.

This is a new version of pull request #23899 which I accidentally closed. The commit `b313907` contains all changes related with the review of @Avogar .

### CODE REVIEW

File: src/AggregateFunctions/AggregateFunctionMap.h [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        const auto & val_column = map_nested_tuple.getColumn(1);
+
+        auto & merged_maps = this->data(place).merged_maps;
+
+        for (size_t i = 0; i < size; ++i)
+        {
+            KeyType key;
+            if constexpr (std::is_same<KeyType, String>::value)
+            {
+                key = key_column.operator[](offset + i).get<KeyType>();
```
Comment (Avogar) [inline src/AggregateFunctions/AggregateFunctionMap.h:97]: It seems like we make double copy of a string here. One on creating Field:
https://github.com/ClickHouse/ClickHouse/blob/38493bde487ce48e37f0fdbcb2894f55d78298c3/src/Core/Field.h#L677
And the second when we perform `operator =` for String. Can we just cast `key_column` to `ColumnString` and use `getData` method and use returned StringRef to construct String?

**Comment by Avogar** [comment]
@ildus How difficult is it to make Map combinator work not only with the Map type, but also with arrays as arguments (like `sumMap`)?

**Comment by ildus** [comment]
@Avogar it's possible if we remove some features of it. The first - `sumMap` supports the input like `[keys array], [values array 1], [values array 2]`, and `[values array 1]` and `[values array 2]` can have different types which requires different nested functions too. But it's not documented, and maybe this feature can be dropped and the combinator will support only one value array.
The second - `sumMap` does compaction of the result value, it removes the keys where corresponding value is the default value (for `int` it's zero). But I think that's somehow can be solved if we pass some temporary column to the nested function, and check that value before putting it to the result column.

**Comment by ildus** [comment]
In the fuzzer I see mostly errors with invalid arguments, how they can be skipped?

**Comment by Avogar** [comment]
Fuzzer faillure is not related to changes: https://github.com/ClickHouse/ClickHouse/issues/24811

File: src/AggregateFunctions/AggregateFunctionMap.h [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    static void readKey(KeyType & key, ReadBuffer & buf) { readBinary(key, buf); }
+};
+
+template <>
+struct AggregateFunctionMapCombinatorData<String>
+{
+    std::unordered_map<String, AggregateDataPtr> merged_maps;
+
+    static void writeKey(String key, WriteBuffer & buf) { writeString(key, buf); }
+    static void readKey(String & key, ReadBuffer & buf) { readString(key, buf); }
```
Comment (Avogar) [inline src/AggregateFunctions/AggregateFunctionMap.h:44]: This is a bad way to serialize/deserialize strings. `readString` reads only until `\n` or `\t`:
https://github.com/ClickHouse/ClickHouse/blob/78f1ecb71f7c4e0b18e848e5c89c09cc49ed185b/src/IO/ReadHelpers.h#L471-L472

See https://github.com/ClickHouse/ClickHouse/blob/170c49db69c18bb18e13b4e55547e26c38b529c9/src/DataTypes/Serializations/SerializationString.cpp#L26
and
https://github.com/ClickHouse/ClickHouse/blob/170c49db69c18bb18e13b4e55547e26c38b529c9/src/DataTypes/Serializations/SerializationString.cpp#L34

File: src/AggregateFunctions/AggregateFunctionMap.h (hunk 1 above)
Comment (ildus) [reply src/AggregateFunctions/AggregateFunctionMap.h:97]: I have some problems with const columns if I use direct casting, investigating

File: src/AggregateFunctions/AggregateFunctionMap.h (hunk 2 above)
Comment (ildus) [reply src/AggregateFunctions/AggregateFunctionMap.h:44]: Good to know, how I can test the serialization?

Comment (Avogar) [reply src/AggregateFunctions/AggregateFunctionMap.h:44]: Serialization/deserialization is used in remote queries. For example:
```sql
SELECT sumMap(map) FROM remote('127.0.0.{1,2,3}', currentDatabase(), table_name);
```

File: src/AggregateFunctions/AggregateFunctionMap.h (hunk 1 above)
Comment (ildus) [reply src/AggregateFunctions/AggregateFunctionMap.h]: Should be fixed, in current version String construction is avoided as much as possible.

**Comment by ildus** [comment]
@Avogar any news on this PR? Tests are passed, failed tests seem unrelated.

**Comment by ildus** [comment]
Cool, thanks @Avogar

**Comment by den-crane** [comment]
@ildus  sumMap has overflow issue with Distributed queries https://github.com/ClickHouse/ClickHouse/issues/35359#issuecomment-1072750645

**Comment by alexey-milovidov** [comment]
@ildus, @Avogar a bug has been found: https://github.com/ClickHouse/ClickHouse/issues/38718

**Comment by alexey-milovidov** [comment]
@ildus Another bug has been found: #40848

**Pull request merged**